### PR TITLE
move to legacy MDG because loom sucks

### PR DESCRIPTION
**Pull request by screret**
## What
switch to the new neoforge Legacy ModDevGradle because loom sucks.

## Implementation Details
mostly copied from 1.21, with help from the neoforge discord for the rest.

## Outcome
no more weird errors hopefully.

### CODE REVIEW

File: dependencies.gradle [hunk 1]
```diff
@@ -40,8 +23,7 @@ dependencies {
     modCompileOnly forge.rei.forge
 
     // EMI
-    modCompileOnly("dev.emi:emi-forge:${forge.versions.emi.get()}:api")
-
+    //modCompileOnly("dev.emi:emi-forge:${forge.versions.emi.get()}:api")
```
Comment (TechLord22) [inline dependencies.gradle:26]: Is commenting this out intentional?

File: gradle/scripts/moddevgradle.gradle [hunk 2]
```diff
@@ -0,0 +1,89 @@
+static def isJetBrainsRuntime() {
```
Comment (TechLord22) [inline gradle/scripts/moddevgradle.gradle]: Unused? Also this isn't reliable, Intellij can be run with the system JDK.

File: .github/actions/build_setup/action.yml [hunk 3]
```diff
@@ -15,11 +15,9 @@ runs:
...
         submodules: true
 
-    - name: Set up JDK 17
-      uses: actions/setup-java@v4
-      with:
-        distribution: adopt
-        java-version: 17
+    - name: Set up JDK 21
+      shell: bash
+      run: echo "JAVA_HOME=$JAVA_HOME_21_X64" >> "$GITHUB_ENV"
```
Comment (TechLord22) [inline .github/actions/build_setup/action.yml]: Why not use setup-java directly for this?

File: .github/actions/build_setup/action.yml [hunk 4]
```diff
@@ -28,9 +26,20 @@ runs:
...
 
+    - name: Cache build directory
+      uses: actions/cache@v4
+      with:
+        path: |
+          build/neoForge
+          build/neoForm
+          build/tmp/.cache/**
+        key: "${{ runner.os }}-gradle-${{ hashFiles('**/libs.versions.*', '**/*.gradle*', '**/gradle-wrapper.properties') }}"
+        restore-keys: "${{ runner.os }}-gradle-"
```
Comment (TechLord22) [inline .github/actions/build_setup/action.yml]: The setup-gradle task should handle this already. I'm not sure why this is separately needed.

File: .github/workflows/auto-build.yml [hunk 5]
```diff
@@ -23,9 +27,7 @@ jobs:
         uses: ./.github/actions/build_setup
 
       - name: Build
-        uses: gradle/gradle-build-action@v2
-        with:
-          arguments: build
+        run: ./gradlew build --build-cache
```
Comment (TechLord22) [inline .github/workflows/auto-build.yml:30]: The Gradle build action will run this with the build cache automatically, and will utilize its own caching. Not sure why this was changed.

File: .github/workflows/auto-publish.yml [hunk 6]
```diff
@@ -36,24 +36,15 @@ jobs:
...
-          distribution: 'adopt'
-
       - name: Setup Build
         uses: ./.github/actions/build_setup
 
       - name: Build
-        uses: gradle/gradle-build-action@v2
-        with:
-          arguments: build
+        run: ./gradlew build --build-cache
```
Comment (TechLord22) [inline .github/workflows/auto-publish.yml:43]: Same as the other similar changes.

File: .github/workflows/build-on-push.yml [hunk 7]
```diff
@@ -32,9 +32,7 @@ jobs:
 
       - name: Build
         if: steps.filter.outputs.code == 'true'
-        uses: gradle/gradle-build-action@v2
-        with:
-          arguments: assemble
+        run: ./gradlew assemble --build-cache
```
Comment (TechLord22) [inline .github/workflows/build-on-push.yml:35]: Same as the other similar changes.

File: dependencies.gradle (hunk 1 above)
Comment (screret) [reply dependencies.gradle:26]: Yep

File: .github/actions/build_setup/action.yml (hunk 4 above)
Comment (screret) [reply .github/actions/build_setup/action.yml]: I read somewhere that this is better, but I'll remove it

File: .github/workflows/auto-build.yml (hunk 5 above)
Comment (screret) [reply .github/workflows/auto-build.yml:30]: gradle build action is deprecated.

**Review by Ghostipedia — APPROVED** [review]

**Pull request merged**